tests/sagediff.py

```diff
@@ -140,70 +135,369 @@ def read_header(self):
-        #     os.lseek(self.fd, offset, os.SEEK_SET)
-        #     tree = os.read(self.fd, nbytes)
-        # 
-        # tree = np.asarray(tree, dtype=self.dtype)
-        
         # This assumes sequential reads
-        tree = np.fromfile(self.fp, dtype=self.dtype, count=ngal)
+        tree = np.fromfile(fp, dtype=self.dtype, count=ngal)
+
+        # If we had to open up the file, close it.
+        if close_file:
+            fp.close()
+
         return tree
 
-def compare_catalogs(g1, g2):
+
+    def update_metadata(self):
+        """
+        The binary galaxies can be split up over multiple files.  In this method, we
+        iterate over the files and collect info that is spread across them.
+        """
+
+        self.totntrees_all_files = 0
+        self.totngals_all_files = 0
+        self.ngal_per_tree_all_files = []
+
+        for file_idx in range(self.num_files):
+
+            # Cut off the number at the end of the file.
+            fname_base = self.filename[:-2]
+
+            # Then append this file number.
+            fname = "{0}_{1}".format(fname_base, file_idx)
+
+            # Open and read the header.
+            fp = open(fname, "rb")
+
+            self.read_header(fp)
+
+            # Then update the global parameters.
+            self.totntrees_all_files += self.totntrees
+            self.totngals_all_files += self.totngals
+            self.ngal_per_tree_all_files.extend(self.ngal_per_tree)
+
+            fp.close() 
+
+
+    def read_gals(self):
+
+        # First determine the number of galaxies across all files.
+        totngals_all_files = 0
+
+        # Initialize an empty array.
+        gals = np.empty(self.totngals_all_files, dtype=self.dtype)
+
+        # Then slice all the galaxies in.
+        offset = 0
+        for file_idx in range(self.num_files):
+
+            fname_base = self.filename[:-2]
+            fname = "{0}_{1}".format(fname_base, file_idx)
+            fp = open(fname, "rb")
+
+            self.read_header(fp)
+            gals_file = self.read_tree(None, fp)
+
+            gals_this_file = len(gals_file)
+
+            gals[offset:offset+gals_this_file] = gals_file
+            offset += gals_this_file 
+
+            fp.close()
+
+        return gals
+
+
+def compare_catalogs(fname1, fname2, mode, num_files, ignored_fields, multidim_fields=None,
+                     rtol=1e-9, atol=5e-5):
     """
     Compares two SAGE catalogs exactly
     """
-    
-    if not (isinstance(g1, sageResults) and 
-            isinstance(g2, sageResults)):
-        msg = "Both inputs must be objects the class 'sageResults'"\
+
+    # For both modes, the first file will be binary.  So lets initialize it.
+    g1 = BinarySage(fname1, ignored_fields)
+
+    # The second file will be either binary or HDF5.
+    if mode == "binary-binary":
+        g2 = BinarySage(fname2, ignored_fields, num_files)
+        compare_binary_catalogs(g1, g2, rtol, atol)
+    else:
```

Comment:
While we do not have a `known-correct` hdf5 output, we might add that in the future. May be replace this `if` with a `switch-case`?